### PR TITLE
Update and merge viscoplasticity models

### PR DESCRIPTION
**Pull request by jhuang2601**
Duplicate and update branch (#2004) for adding viscoplasticity (Duvaut-Lions model)

No rebaseline is required for this PR

As discussed,  ViscoDelftEgg model, validation examples and integrated tests with boundary value problems will be added in  separate PRs.

### CODE REVIEW

**Review by sytuannguyen — APPROVED** [review]
Great! I verified that all the tests work perfectly. Thank you @jhuang2601

**Review by herve-gross — APPROVED** [review]
Great work to all! One day (not this PR), we should go systematically in the code and documentation and do a search/replace to change GEOSX->GEOS.

File: src/coreComponents/physicsSolvers/multiphysics/poromechanicsKernels/PoromechanicsKernels.cmake [hunk 1]
```diff
@@ -21,7 +21,10 @@ set( porousSolidDispatch PorousSolid<DruckerPragerExtended>
                          PorousSolid<ElasticOrthotropic>
                          PorousSolid<DamageSpectral<ElasticIsotropic>>
                          PorousSolid<DamageVolDev<ElasticIsotropic>>
-                         PorousSolid<Damage<ElasticIsotropic>> )
+                         PorousSolid<Damage<ElasticIsotropic>>
+                         PorousSolid<DuvautLionsSolid<DruckerPrager>>
+                         PorousSolid<DuvautLionsSolid<DruckerPragerExtended>>
+                         PorousSolid<DuvautLionsSolid<ModifiedCamClay>> )
```
Comment (francoishamon) [inline src/coreComponents/physicsSolvers/multiphysics/poromechanicsKernels/PoromechanicsKernels.cmake:27]: Before merging, we should find a way to pass the CI test by reducing the number of options here: https://fcmaelstrom.slack.com/archives/C03U0RPR5K2/p1678918573857919?thread_ts=1678914869.324949&cid=C03U0RPR5K2

File: src/coreComponents/physicsSolvers/multiphysics/poromechanicsKernels/SinglePhasePoromechanics_impl.hpp [hunk 2]
```diff
@@ -85,11 +85,13 @@ smallStrainUpdate( localIndex const k,
   real64 dPorosity_dPressure = 0.0;
   real64 dPorosity_dTemperature = 0.0;
   real64 dSolidDensity_dPressure = 0.0;
+  real64 timeIncrement = 0.0;
```
Comment (francoishamon) [inline src/coreComponents/physicsSolvers/multiphysics/poromechanicsKernels/SinglePhasePoromechanics_impl.hpp:88]: Here and in the other mechanics/poromechanics kernels, should you pass the actual `dt` to have a complete implementation? Currently the `timeIncrement` are all hard-coded to 0.

File: src/coreComponents/constitutive/solid/PorousSolid.hpp [hunk 3]
```diff
@@ -83,11 +84,12 @@ class PorousSolidUpdates : public CoupledSolidUpdates< SOLID_TYPE, BiotPorosity,
                                        real64 & dPorosity_dTemperature,
                                        real64 & dSolidDensity_dPressure ) const
   {
-
+    GEOSX_UNUSED_VAR( timeIncrement );
```
Comment (francoishamon) [inline src/coreComponents/constitutive/solid/PorousSolid.hpp:87]:
```suggestion
```

File: src/coreComponents/constitutive/solid/DuvautLionsSolid.hpp [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+{
+public:
+  template< typename ... PARAMS >
+  DuvautLionsSolidUpdates( real64 const & relaxationTime,
+                           PARAMS && ... baseParams ):
+    UPDATE_BASE( std::forward< PARAMS >( baseParams )... ),
+    m_relaxationTime( relaxationTime )
+  {}
+
+  // using DiscretizationOps = typename UPDATE_BASE::DiscretizationOps;
```
Comment (francoishamon) [inline src/coreComponents/constitutive/solid/DuvautLionsSolid.hpp:58]:
```suggestion
```

File: src/coreComponents/constitutive/solid/DuvautLionsSolid.hpp [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+      trialStress[i] = stress[i];
+    }
+
+    if( m_disableInelasticity )
+    {
+      return;
+    }
+
+    //Get trial stress and elastic stiffness by disabling inelasticity
+    // m_solidUpdate.m_disableInelasticity = true;
```
Comment (francoishamon) [inline src/coreComponents/constitutive/solid/DuvautLionsSolid.hpp:92]: Do you still need to disable inelasticity here? Or `smallStrainUpdate_ElasticOnly` will do it automatically? If it will, you can remove the commented code here and below.

File: src/coreComponents/constitutive/solid/ElasticIsotropic.hpp [hunk 6]
```diff
@@ -287,11 +291,13 @@ GEOSX_HOST_DEVICE
 GEOSX_FORCE_INLINE
 void ElasticIsotropicUpdates::smallStrainUpdate( localIndex const k,
                                                  localIndex const q,
+                                                 real64 const & timeIncrement,
                                                  real64 const ( &strainIncrement )[6],
                                                  real64 ( & stress )[6],
                                                  real64 ( & stiffness )[6][6] ) const
 {
-  smallStrainUpdate_StressOnly( k, q, strainIncrement, stress );
+  GEOSX_UNUSED_VAR( timeIncrement );
```
Comment (francoishamon) [inline src/coreComponents/constitutive/solid/ElasticIsotropic.hpp:299]:
```suggestion
```

File: src/coreComponents/constitutive/solid/DelftEgg.hpp [hunk 7]
```diff
@@ -210,13 +212,14 @@ GEOSX_HOST_DEVICE
...
                                          localIndex const q,
+                                         real64 const & timeIncrement,
                                          real64 const ( &strainIncrement )[6],
                                          real64 ( & stress )[6],
                                          real64 ( & stiffness )[6][6] ) const
 {
 
   // Rename variables for easier implementation
-
+  GEOSX_UNUSED_VAR( timeIncrement );
```
Comment (francoishamon) [inline src/coreComponents/constitutive/solid/DelftEgg.hpp:222]:
```suggestion
```

File: src/coreComponents/constitutive/solid/DruckerPrager.hpp [hunk 8]
```diff
@@ -141,13 +160,14 @@ GEOSX_HOST_DEVICE
...
                                               localIndex const q,
+                                              real64 const & timeIncrement,
                                               real64 const ( &strainIncrement )[6],
                                               real64 ( & stress )[6],
                                               real64 ( & stiffness )[6][6] ) const
 {
   // elastic predictor (assume strainIncrement is all elastic)
-
-  ElasticIsotropicUpdates::smallStrainUpdate( k, q, strainIncrement, stress, stiffness );
+  GEOSX_UNUSED_VAR( timeIncrement );
```
Comment (francoishamon) [inline src/coreComponents/constitutive/solid/DruckerPrager.hpp:169]:
```suggestion
```

File: src/coreComponents/constitutive/solid/DruckerPrager.hpp [hunk 9]
```diff
@@ -293,16 +313,31 @@ void DruckerPragerUpdates::smallStrainUpdate( localIndex const k,
...
+GEOSX_FORCE_INLINE
+void DruckerPragerUpdates::smallStrainUpdate_ElasticOnly( localIndex const k,
+                                                          localIndex const q,
+                                                          real64 const & timeIncrement,
+                                                          real64 const ( &strainIncrement )[6],
+                                                          real64 ( & stress )[6],
+                                                          real64 ( & stiffness )[6][6] ) const
+{
+  // elastic predictor (assume strainIncrement is all elastic)
+  GEOSX_UNUSED_VAR( timeIncrement );
```
Comment (francoishamon) [inline src/coreComponents/constitutive/solid/DruckerPrager.hpp:326]:
```suggestion
```

File: src/coreComponents/constitutive/solid/ElasticIsotropic.hpp [hunk 10]
```diff
@@ -300,15 +306,28 @@ GEOSX_HOST_DEVICE
 GEOSX_FORCE_INLINE
 void ElasticIsotropicUpdates::smallStrainUpdate( localIndex const k,
                                                  localIndex const q,
+                                                 real64 const & timeIncrement,
                                                  real64 const ( &strainIncrement )[6],
                                                  real64 ( & stress )[6],
                                                  DiscretizationOps & stiffness ) const
 {
-  smallStrainUpdate_StressOnly( k, q, strainIncrement, stress );
+  GEOSX_UNUSED_VAR( timeIncrement );
```
Comment (francoishamon) [inline src/coreComponents/constitutive/solid/ElasticIsotropic.hpp:314]:
```suggestion
```

File: src/coreComponents/constitutive/solid/ElasticOrthotropic.hpp [hunk 11]
```diff
@@ -298,11 +306,13 @@ GEOSX_HOST_DEVICE
 GEOSX_FORCE_INLINE
 void ElasticOrthotropicUpdates::smallStrainUpdate( localIndex const k,
                                                    localIndex const q,
+                                                   real64 const & timeIncrement,
                                                    real64 const ( &strainIncrement )[6],
                                                    real64 ( & stress )[6],
                                                    DiscretizationOps & stiffness ) const
 {
-  smallStrainUpdate_StressOnly( k, q, strainIncrement, stress );
+  GEOSX_UNUSED_VAR( timeIncrement );
```
Comment (francoishamon) [inline src/coreComponents/constitutive/solid/ElasticOrthotropic.hpp:314]:
```suggestion
```

File: src/coreComponents/constitutive/solid/ElasticOrthotropic.hpp [hunk 12]
```diff
@@ -285,11 +291,13 @@ GEOSX_FORCE_INLINE
 GEOSX_HOST_DEVICE
 void ElasticOrthotropicUpdates::smallStrainUpdate( localIndex const k,
                                                    localIndex const q,
+                                                   real64 const & timeIncrement,
                                                    real64 const ( &strainIncrement )[6],
                                                    real64 ( & stress )[6],
                                                    real64 ( & stiffness )[6][6] ) const
 {
-  smallStrainUpdate_StressOnly( k, q, strainIncrement, stress );
+  GEOSX_UNUSED_VAR( timeIncrement );
```
Comment (francoishamon) [inline src/coreComponents/constitutive/solid/ElasticOrthotropic.hpp:299]:
```suggestion
```

File: src/coreComponents/constitutive/solid/ElasticTransverseIsotropic.hpp [hunk 13]
```diff
@@ -257,23 +263,27 @@ GEOSX_FORCE_INLINE
 GEOSX_HOST_DEVICE
 void ElasticTransverseIsotropicUpdates::smallStrainUpdate( localIndex const k,
                                                            localIndex const q,
+                                                           real64 const & timeIncrement,
                                                            real64 const ( &strainIncrement )[6],
                                                            real64 ( & stress )[6],
                                                            real64 ( & stiffness )[6][6] ) const
 {
-  smallStrainUpdate_StressOnly( k, q, strainIncrement, stress );
+  GEOSX_UNUSED_VAR( timeIncrement );
```
Comment (francoishamon) [inline src/coreComponents/constitutive/solid/ElasticTransverseIsotropic.hpp:271]:
```suggestion
```

File: src/coreComponents/constitutive/solid/ElasticTransverseIsotropic.hpp [hunk 14]
```diff
@@ -257,23 +263,27 @@ GEOSX_FORCE_INLINE
...
 GEOSX_FORCE_INLINE
 void ElasticTransverseIsotropicUpdates::smallStrainUpdate( localIndex const k,
                                                            localIndex const q,
+                                                           real64 const & timeIncrement,
                                                            real64 const ( &strainIncrement )[6],
                                                            real64 ( & stress )[6],
                                                            DiscretizationOps & stiffness ) const
 {
-  smallStrainUpdate_StressOnly( k, q, strainIncrement, stress );
+  GEOSX_UNUSED_VAR( timeIncrement );
```
Comment (francoishamon) [inline src/coreComponents/constitutive/solid/ElasticTransverseIsotropic.hpp:285]:
```suggestion
```

File: src/coreComponents/physicsSolvers/solidMechanics/kernels/ExplicitSmallStrain_impl.hpp [hunk 15]
```diff
@@ -129,11 +130,12 @@ void ExplicitSmallStrain< SUBREGION_TYPE, CONSTITUTIVE_TYPE, FE_TYPE >::quadratu
   real64 const detJ = FE_TYPE::inverseJacobianTransformation( q, stack.xLocal, invJ );
 
   real64 strain[6] = {0};
+  real64 timeIncrement;
```
Comment (francoishamon) [inline src/coreComponents/physicsSolvers/solidMechanics/kernels/ExplicitSmallStrain_impl.hpp:133]: I would suggest to initialize this variable to avoid issues later

**Review by francoishamon — APPROVED** [review]
Looks great, I just made some tiny cosmetic comments

File: src/coreComponents/physicsSolvers/multiphysics/poromechanicsKernels/PoromechanicsKernels.cmake (hunk 1 above)
Comment (jhuang2601) [reply src/coreComponents/physicsSolvers/multiphysics/poromechanicsKernels/PoromechanicsKernels.cmake:27]: Right, this becomes an obvious hurdle on Travis build

File: src/coreComponents/physicsSolvers/multiphysics/poromechanicsKernels/SinglePhasePoromechanics_impl.hpp (hunk 2 above)
Comment (jhuang2601) [reply src/coreComponents/physicsSolvers/multiphysics/poromechanicsKernels/SinglePhasePoromechanics_impl.hpp:88]: This will be fixed in the following PR, when working on boundary value problems.

File: src/coreComponents/constitutive/solid/PorousSolid.hpp (hunk 3 above)
Comment (jhuang2601) [reply src/coreComponents/constitutive/solid/PorousSolid.hpp]: Done

File: src/coreComponents/constitutive/solid/DuvautLionsSolid.hpp (hunk 5 above)
Comment (jhuang2601) [reply src/coreComponents/constitutive/solid/DuvautLionsSolid.hpp]: Done

File: src/coreComponents/constitutive/solid/DelftEgg.hpp (hunk 7 above)
Comment (jhuang2601) [reply src/coreComponents/constitutive/solid/DelftEgg.hpp]: Done

File: src/coreComponents/constitutive/solid/DruckerPrager.hpp (hunk 8 above)
Comment (jhuang2601) [reply src/coreComponents/constitutive/solid/DruckerPrager.hpp]: Done

File: src/coreComponents/constitutive/solid/DruckerPrager.hpp (hunk 9 above)
Comment (jhuang2601) [reply src/coreComponents/constitutive/solid/DruckerPrager.hpp]: Done

File: src/coreComponents/constitutive/solid/ElasticIsotropic.hpp (hunk 10 above)
Comment (jhuang2601) [reply src/coreComponents/constitutive/solid/ElasticIsotropic.hpp]: Done

File: src/coreComponents/constitutive/solid/ElasticOrthotropic.hpp (hunk 11 above)
Comment (jhuang2601) [reply src/coreComponents/constitutive/solid/ElasticOrthotropic.hpp]: Done

File: src/coreComponents/constitutive/solid/ElasticOrthotropic.hpp (hunk 12 above)
Comment (jhuang2601) [reply src/coreComponents/constitutive/solid/ElasticOrthotropic.hpp]: Done

File: src/coreComponents/constitutive/solid/ElasticTransverseIsotropic.hpp (hunk 13 above)
Comment (jhuang2601) [reply src/coreComponents/constitutive/solid/ElasticTransverseIsotropic.hpp]: Done

File: src/coreComponents/constitutive/solid/ElasticTransverseIsotropic.hpp (hunk 14 above)
Comment (jhuang2601) [reply src/coreComponents/constitutive/solid/ElasticTransverseIsotropic.hpp]: Done

File: src/coreComponents/physicsSolvers/solidMechanics/kernels/ExplicitSmallStrain_impl.hpp (hunk 15 above)
Comment (jhuang2601) [reply src/coreComponents/physicsSolvers/solidMechanics/kernels/ExplicitSmallStrain_impl.hpp]: You're right.

File: src/coreComponents/physicsSolvers/multiphysics/poromechanicsKernels/PoromechanicsKernels.cmake (hunk 1 above)
Comment (joshua-white) [reply src/coreComponents/physicsSolvers/multiphysics/poromechanicsKernels/PoromechanicsKernels.cmake:27]: @Jian, let's just comment out these instances for the moment (anything with DuvautLionSolid) until we resolve the build time issues.

**Comment by francoishamon** [comment]
I am going to merge this PR since we can now pass the CI tests and all integrated tests are passing as well.
To summarize the latest developments made to pass CI tests:
- The new viscoplastic model is not available for the solid mechanics and poromechanics kernels. The viscoplastic model is not included in the [constitutive dispatch](https://github.com/GEOS-DEV/GEOS/blob/8865896b9df11a7753aefdf4a38b451ec7935343/src/coreComponents/constitutive/ConstitutivePassThru.hpp#L84) for these solvers.
- The new viscoplastic model is only available through the triaxial driver thanks to a [separate constitutive dispatch](https://github.com/GEOS-DEV/GEOS/blob/8865896b9df11a7753aefdf4a38b451ec7935343/src/coreComponents/constitutive/ConstitutivePassThru.hpp#L123) that will need to be removed at some point.
- We started removing some `virtual` and `override` keywords for the `smallStrainUpdate` calls in the solid models, but many are remaining and a larger refactoring of the solid model is needed to clean this up.
